tests/benchmark/utils/s2n_server_recv_benchmark.cc

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tests/benchmark/utils/s2n_server_recv_benchmark.h"
+#include "tests/benchmark/utils/shared_info.h"
+#include "utils/s2n_safety.h"
+#include <unistd.h>
+#include <stdio.h>
+#include <vector>
+#include <stdlib.h>
+#include <poll.h>
+
+extern "C" {
+#include "bin/common.h"
+#include "tls/s2n_connection.h"
+}
+
+int recv_data(struct s2n_connection *conn, int sockfd2, benchmark::State& state, bool warmup) {
+    struct pollfd readers[1];
+    readers[0].fd = sockfd2;
+    readers[0].events = POLLIN;
+    errno = 0;
+    int poll_output = 0;
+    s2n_blocked_status blocked;
+    uint8_t* recieved;
+    recieved = (uint8_t*)malloc(state.range(0) + 2);
+
+    while ((poll_output = poll(readers, 1, -1)) > 0) {
+        int bytes_read = 0;
+        do {
+            int received_bytes = 0;
+            if (!warmup) {
+                state.ResumeTiming();
+            }
+            benchmark::DoNotOptimize(received_bytes = s2n_recv(conn, recieved + bytes_read, (state.range(0) + 2) - bytes_read, &blocked));
+            if (!warmup) {
+                state.PauseTiming();
+            }
+            if (received_bytes < 0) {
+                fprintf(stderr, "Error reading from connection: '%s' %d\n", s2n_strerror(s2n_errno, "EN"),
+                        s2n_connection_get_alert(conn));
+                exit(1);
+            }
+            bytes_read += received_bytes;
+        } while (bytes_read < state.range(0) + 2);
+        break;
+    }
+    free(recieved);
+    return 0;
+}
+
+static int benchmark_recv_single_suite_server(benchmark::State& state) {
+    int connectionfd = state.range(3);
+    struct s2n_connection *conn = s2n_connection_new(S2N_SERVER);
+    size_t warmup_iters = state.range(2);
+
+    if (!conn) {
+        print_s2n_error("Error getting new s2n connection");
+        exit(1);
+    }
+
+    s2n_setup_server_connection(conn, connectionfd, config, conn_settings);
+
+    GUARD_EXIT(benchmark_negotiate(conn, connectionfd, state, true), "Server negotiation failed\n");
+    if (conn_settings.mutual_auth) {
+        if (!s2n_connection_client_cert_used(conn)) {
+            print_s2n_error("Error: Mutual Auth was required, but not negotiated");
+            exit(1);
+        }
+    }
```

Comment:
1. I like the reuse here: using the negotiate code to setup the connection for the send/recv benchmark
2. Can even more be shared? Maybe another option to one unified client/server benchmark that can handle both the negotiate and send/recv